Refactor the DB to use the schema discussed in issue #2

denheck opened the pull request.

denheck committed:
initial update of question fetching mix task

denheck committed:
refactor mix task used to fetch questions

denheck committed:
refactor generate fake answers

alexcavalli commented:
@denheck Soooo looking at this more, I kinda wonder if we're overcomplicating it still (which would certainly be my fault for overcomplicating it in the first place). 

I think the way I had it originally, with trivia_items as a table that included all the info from a wikipedia article, sort of made the table structure make sense, but the table structure was pretty useless and built way more around the data being ingested and not enough around the data as it was being used.

I think your change to use `questions` as the questions is a good change (I mean obviously), but I'm kinda wondering now if we might want to avoid all the shenanigans with answers and just make questions an "ETL" kind of result from trivia_items. I.e. tables like this:

```
trivia_items:
  title:
  description:
  redacted_description: (this could arguably not exist here)

questions:
  question: (the redacted description)
  answer_choices: ["John", "Puerto Rico", "etc"] (as a serialized array or something)
  correct_answer: "Puerto Rico"
```
With optionally a link between the question and the trivia item that it references (i.e. a trivia_item_id for the correct answer, but that should be equivalent to the trivia item with a title = the correct answer). And then a load process like this:

1. Pull down wikipedia data as trivia items
2. Generate complete questions, totally independent from trivia item data, from the trivia item data.

denheck committed:
add JSON to mix.exs

denheck committed:
update .gitignore

denheck committed:
loading questions into JSON file

denheck committed:
simplify db

denheck committed:
rename answers to answer_choices

denheck committed:
remove unused answer model

denheck committed:
update question fetching and generating

denheck commented:
@alexcavalli I refactored `mix wikitriva.fetch_questions` to fetch the set of questions from from Wikipedia and load them into a temp file at `data/questions.json` to avoid the need for a `trivia_items` table. Also, I updated the schema of `questions` to match what you mentioned above and refactored `mix wikitrivia.generate_questions` to transform this file into a set of questions which can be loaded into the `questions` table.

This simplified the mix tasks quite a bit and properly separated the fetching from the transform/import steps. It's possible we may want to compose these into a new mix task that does both the fetching and transforming/importing at some point.

alexcavalli commented on lib/wikitrivia/question.ex:
```diff
@@ -3,17 +3,16 @@ defmodule Wikitrivia.Question do
-    many_to_many :answer_choices, Wikitrivia.TriviaItem, join_through: "questions_trivia_items"
+    field :question, :string
+    field :correct_answer, :string
+    field :answer_choices, {:array, :string}
```
@denheck Very cool that this also just worked.

denheck replied:
yeah, it really simplified the DB structure. Glad you mentioned that Postgres had an array column type

alexcavalli commented on .gitignore:
```diff
@@ -25,3 +25,4 @@ erl_crash.dump
 .vscode
 .elixir_ls
 assets/node_modules
+data/*
```
Maybe we should rename this to `s3/`? :P

denheck replied:
🤣 the old fashioned local s3

alexcavalli committed:
Minor cleanup to question generation mix tasks

alexcavalli reviewed (approved):
@denheck LGTM. I made a couple small changes (removed some dead code and fixed a minor bug). 

Definitely agree that we could simplify these into a single task (they're pretty dependent) but I think this works for our needs as-is.

denheck commented on lib/mix/tasks/wikitrivia.fetch_questions.ex:
```diff
@@ -28,6 +28,7 @@ defmodule Mix.Tasks.Wikitrivia.FetchQuestions do
   end
 
   defp load_questions(questions) do
+    File.mkdir_p("data")
```
👍
